cachi2/core/package_managers/yarn/locators.py

```diff
@@ -1,10 +1,208 @@
 import re
 from dataclasses import dataclass
 from functools import cached_property
-from typing import NamedTuple, Optional
+from pathlib import Path
+from typing import NamedTuple, Optional, Sequence, Union
 from urllib.parse import parse_qs, unquote
 
-from cachi2.core.errors import UnexpectedFormat
+from cachi2.core.errors import UnexpectedFormat, UnsupportedFeature
+
+# https://github.com/yarnpkg/berry/blob/b6026842dfec4b012571b5982bb74420c7682a73/packages/plugin-http/sources/constants.ts
+TARBALL_RE = re.compile(r"^[^?]*\.(?:tar\.gz|tgz)(?:\?.*)?(?:#.*)?$")
+
+# --- Locator types ---
+
+
+class NpmLocator(NamedTuple):
+    """Locator that handles registry dependencies.
+
+    Sample locator string:
+        [@scope/]name@npm:version
+
+    Attributes:
+        scope: the scope of the dependency (without '@')
+        name: the name of the dependency
+        version: the semver version of the dependency
+    """
+
+    scope: Optional[str]
+    name: str
+    version: str
+
+
+class WorkspaceLocator(NamedTuple):
+    """Locator that handles workspace dependencies.
+
+    Sample locator string:
+        [@scope/]name@workspace:path/to/dir
+
+    Attributes:
+        scope: the scope of the dependency (without '@')
+        name: the name of the dependency
+        relpath: relative path from the project root to the workspace
+    """
+
+    # The scope and name in a workspace locator seem reliable, yarnberry won't let you use an
+    # arbitrary name (it must match the workspace's package.json)
+    scope: Optional[str]
+    name: str
+    relpath: Path
+
+
+class PatchLocator(NamedTuple):
+    """Locator that handles patched dependencies.
+
+    Sample locator string:
+        [@scope/]name@patch:<patched-dep-locator>#path/to/patch::version=1.0.0&hash=abc123&locator=<workspace-locator>
+        [@scope/]name@patch:<patched-dep-locator>#~builtin<compat/patch-name>::version=1.0.0&hash=abc123
+
+    Attributes:
+        package: the dependency that gets patched
+        patches: a list of paths (relative to a parent locator) or builtin patch identifiers
+            (e.g. "builtin<compat/typescript>")
+        locator: relative paths (if any) are relative to this locator
+    """
+
+    # The scope and name in a patch locator (and most locators other than Npm and Workspace) are
+    # not reliable. Yarnberry *will* let you use a completely arbitrary name.
+
+    package: "Locator"
+    patches: Sequence[Union[str, Path]]
+    locator: Optional["Locator"]
```

Comment:
Question: do you want to raise UnsupportedFeature errors based on the Locator type? If so, when?

* What will you do if the `package` is a `PatchLocator`?
* What will you do if the `locator` is anything other than a `WorkspaceLocator`?

Leave it for the SBOM generation?

Reply:
```shell
$ yarn patch 'left-pad@patch:left-pad@npm%3A1.3.0#./my-patches/left-pad.patch::version=1.3.0&hash=629bda&locator=berryscary%40workspace%3A.'
$ yarn patch-commit -s /tmp/xfs-1654aa57/user                                                                                               
Internal Error: Assertion failed: Expected the resolution to have been registered
    at Ag.execute (/home/acmiel/RedHat/Cachito/cachito-testing/cachi2-yarn-berry/.yarn/releases/yarn-3.6.1.cjs:742:4893)
    at async Ag.validateAndExecute (/home/acmiel/RedHat/Cachito/cachito-testing/cachi2-yarn-berry/.yarn/releases/yarn-3.6.1.cjs:345:664)
    at async Gn.run (/home/acmiel/RedHat/Cachito/cachito-testing/cachi2-yarn-berry/.yarn/releases/yarn-3.6.1.cjs:359:2057)
    at async Gn.runExit (/home/acmiel/RedHat/Cachito/cachito-testing/cachi2-yarn-berry/.yarn/releases/yarn-3.6.1.cjs:359:2241)
    at async i (/home/acmiel/RedHat/Cachito/cachito-testing/cachi2-yarn-berry/.yarn/releases/yarn-3.6.1.cjs:446:12054)
    at async t (/home/acmiel/RedHat/Cachito/cachito-testing/cachi2-yarn-berry/.yarn/releases/yarn-3.6.1.cjs:446:10283)
```
> Does it even work?

No :sweat_smile:

Reply:
So `package` can be any supported locator besides `PatchLocator`?

And `locator` needs to be `[Workspace|File|Https]Locator`? I'm assuming all of those work with Yarn, and that `NpmLocator` doesn't. A `PatchLocator` `locator` in a `PatchLocator` seems redundant :sweat_smile:

Reply:
> So package can be any supported locator besides PatchLocator?

Npm and Git locators work for sure, we don't test any others but they would probably work

> And locator needs to be [Workspace|File|Https]Locator? I'm assuming all of those work with Yarn, and that NpmLocator doesn't. A PatchLocator locator in a PatchLocator seems redundant 😅

Yarnberry uses the configured Fetcher to fetch the parent locator https://github.com/yarnpkg/berry/blob/b6026842dfec4b012571b5982bb74420c7682a73/packages/plugin-patch/sources/patchUtils.ts#L124-L129
In other words, I have no idea which ones are supported but I would assume all of them

Reply:
Git locators as `locator` do work at least, and since there are no explicit restrictions others probably do as well
```yaml
"left-pad@patch:left-pad@npm%3A1.3.0#./my-patches/left-pad.patch::version=1.3.0&hash=629bda&locator=berryscary%40https%3A%2F%2Fgithub.com%2Fcachito-testing%2Fcachi2-yarn-berry.git%23commit%3D452d7f530513206de63af402536e9ff736c2aa79":
  version: 1.3.0
  resolution: "left-pad@patch:left-pad@npm%3A1.3.0#./my-patches/left-pad.patch::version=1.3.0&hash=629bda&locator=berryscary%40https%3A%2F%2Fgithub.com%2Fcachito-testing%2Fcachi2-yarn-berry.git%23commit%3D452d7f530513206de63af402536e9ff736c2aa79"
  checksum: 1fb795aff3df86466fa2063ce1db288da15de0db63c624a3d921abab6b3d079093b035c093382fc009afc07da33d65c97b6728ded0efedac999f6f6e1ddf13c3
  languageName: node
  linkType: hard
```

Reply:
That said, the CLI workflow for creating a patch always uses a workspace locator

Reply:
Well, after fiddling a little with patches and reading the docs, I really think they're intended to be local files. Even if you mess with the "resolved" content in `package.json`, it always tries to resolve to a local path.

So I believe that we can ban all patches that are not workspaces, and we can improve if we figure out later that there are corner cases. I guess it's something that will rarely be used, and when it is, it's probably by using the standard CLI workflow.

Reply:
Ack, added the extra checks